### ISSUE TITLE
java.sql.SQLNonTransientConnectionException with MariaDB 10.2.12

### ISSUE THREAD

**Issue by AntonTsaberyaby** [metabase/metabase#10063]
Very often I see such errors in the Metabase logs:

```
05-30 05:02:20 ERROR middleware.log :: GET /api/database 500 5 ms (1 DB calls)
{:message "(conn=505) Broken pipe (Write failed)",
 :type java.sql.SQLNonTransientConnectionException,
 :stacktrace
 ("org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:234)"
  "org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:165)"
  "org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:238)"
  "org.mariadb.jdbc.MariaDbPreparedStatementClient.executeInternal(MariaDbPreparedStatementClient.java:232)"
  "org.mariadb.jdbc.MariaDbPreparedStatementClient.execute(MariaDbPreparedStatementClient.java:159)"
  "org.mariadb.jdbc.MariaDbPreparedStatementClient.executeQuery(MariaDbPreparedStatementClient.java:174)"
  "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:379)"
  "clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1072)"
  "clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1066)"
  "clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1095)"
  "clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1075)"
  "clojure.java.jdbc$query.invokeStatic(jdbc.clj:1164)"
  "clojure.java.jdbc$query.invoke(jdbc.clj:1126)"
  "toucan.db$query.invokeStatic(db.clj:274)"
  "toucan.db$query.doInvoke(db.clj:270)"
  "clojure.lang.RestFn.invoke(RestFn.java:410)"
  "toucan.db$simple_select.invokeStatic(db.clj:378)"
  "toucan.db$simple_select.invoke(db.clj:367)"
  "toucan.db$select.invokeStatic(db.clj:646)"
  "toucan.db$select.doInvoke(db.clj:640)"
  "clojure.lang.RestFn.invoke(RestFn.java:423)"
  "--> api.database$dbs_list.invokeStatic(database.clj:143)"
  "api.database$dbs_list.invoke(database.clj:142)"
  "api.database$fn__47932.invokeStatic(database.clj:155)"
  "api.database$fn__47932.invoke(database.clj:148)"
  "middleware.auth$enforce_authentication$fn__59926.invoke(auth.clj:14)"
  "routes$fn__61084$fn__61085.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__60017.invoke(exceptions.clj:104)"
  "middleware.exceptions$catch_api_exceptions$fn__60014.invoke(exceptions.clj:92)"
  "middleware.log$log_api_call$fn__61458$fn__61459.invoke(log.clj:170)"
  "middleware.log$log_api_call$fn__61458.invoke(log.clj:164)"
  "middleware.security$add_security_headers$fn__59980.invoke(security.clj:105)"
  "middleware.json$wrap_json_body$fn__61163.invoke(json.clj:61)"
  "middleware.json$wrap_streamed_json_response$fn__61181.invoke(json.clj:97)"
  "middleware.session$bind_current_user$fn__58593$fn__58594.invoke(session.clj:193)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:176)"
  "middleware.session$do_with_current_user.invoke(session.clj:170)"
  "middleware.session$bind_current_user$fn__58593.invoke(session.clj:192)"
  "middleware.session$wrap_current_user_id$fn__58582.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__58567.invoke(session.clj:123)"
  "middleware.auth$wrap_api_key$fn__59934.invoke(auth.clj:27)"
  "middleware.misc$maybe_set_site_url$fn__61523.invoke(misc.clj:58)"
  "middleware.misc$bind_user_locale$fn__61526.invoke(misc.clj:74)"
  "middleware.misc$add_content_type$fn__61511.invoke(misc.clj:30)"
  "middleware.misc$disable_streaming_buffering$fn__61548.invoke(misc.clj:119)"
  "middleware.misc$wrap_gzip$fn__61541.invoke(misc.clj:101)"),
 :sql-exception-chain ["SQLNonTransientConnectionException:" "Message: (conn=505) Broken pipe (Write failed)" "SQLState: 08" "Error Code: 0"]}
```

Database logs:

`2019-05-30  5:12:20 140225039922944 [Warning] Aborted connection 501 to db: 'metabase' user: 'root' host: '192.168.224.3' (Got timeout reading communication packets)`

- operating system: Linux
- databases: MariaDB 10.2.12
- Metabase version: 0.32.8
- Metabase hosting environment: Docker
- Metabase internal database: Mysql

Have you any idea how Can I fix this?

**Comment by flamber** [comment]
Hi @AntonTsaberyaby 
When you get that error log, do you then see the message "There was a problem displaying this chart" display on the dashboard cards? And if you do a browser refresh, then it works?
If yes, then I think you're seeing issue #9989

**Comment by AntonTsaberyaby** [comment]
@flamber No, at browser I see error 
`(conn=484) unexpected end of stream, read 0 bytes from 4 (socket was closed by server)`

**Comment by flamber** [comment]
Okay, then you're probably seeing issue #9885

**Comment by pulponair** [comment]
> @flamber No, at browser I see error
> `(conn=484) unexpected end of stream, read 0 bytes from 4 (socket was closed by server)`

That usually happens when:
JDBCs maxIdleTime is higher than marias/mysql wait_timeout.
maxIdleTime of the mariaDB J/Connector defaults to 600s (10 minutes). 

Check your servers wait_timeout by issuing:
"show variables like 'wait_timeout';"

And within Addtional JDBC Connection Options adjust it to something like:
`maxIdleTime=60&sessionVariables=wait_timeout=180`

maxIdleTime always has to be lower than wait_timeout.

That also applies to #9885 and #9989 
That Problem was introduced by switching from MySQL to mariaDB Connector.

**Comment by AntonTsaberyaby** [comment]
@pulponair 

```
MariaDB [(none)]> show variables like 'wait_timeout';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| wait_timeout  | 28800 |
+---------------+-------+
1 row in set (0.00 sec)

MariaDB [(none)]>
```
This default value, I haven't change it anymore

How Can I look at JDBCs maxIdleTime value into my instance?

**Comment by pulponair** [comment]
Hi,
> @pulponair
> How Can I look at JDBCs maxIdleTime value into my instance?
It's 600 by default see:
![image](https://user-images.githubusercontent.com/289596/58943201-8ac0d380-877f-11e9-9809-cdae143d238b.png)
or:
![image](https://user-images.githubusercontent.com/289596/58943280-b5129100-877f-11e9-8a87-c87bd4de2bd9.png)

So your wait_timeout is ok there.

I have several databases suffering on the same issue where wait_timeout is higher then maxIdleTime and i wasnt able to finde a solution yet for them.

Do you have full access to the the SQL server? 
If so:
- on the client side set: maxIdleTime=60&sessionVariables=wait_timeout=180 
- Increase log level to maximum on the server side
- Open a random report beloging to that database in Mebase
- Don't do anything in Metabase that touches the Database
- Obeserve the Serverlogs for ~180s and see what happens there.
- After 180s try to open the same report (you should see a brocken pipe/stream error there).

Guess:
Server is ending the Connection and Client does not notice and runs into a broken pipe
That's eighter related to:
-   a shitty mariaDB j/connector (People on the net sys that older versions work much better but i did not try jet) 
- Metabase/Driver not reconnecting transparently
- Some mix settings between server/client (though i tried almost every possible combination).

If i find time on the weekend i am gonna try to build Metabase using and older MariaDB driver then the one which comes with the current version and report back here.

**Comment by pulponair** [comment]
I tried a couple of MariaDB Driver versions:

2.4.1 -> same Problem
2.3.0 (original one) -> same Problem
2.2.6 -> same Problem
1.8.0 -> same Problem
1.7.3 -> same Problem
1.6.4 -> same Problem
1.5.9 -> same Problem
1.4.6 -> same Problem
1.3.7 -> same Problem
1.2.3 -> same Problem
1.1.9 -> same Problem

So this is either a settings issue or a Metabase Bug.
After all on can say if wait_timeout is reached without any activity on the connection the DB closes the Socket and Metabase is not noticing that and sending queries to a dead socket which results in :

"could not read resultset: unexpected end of stream, read 0 bytes from 7"

Reproducable by:
1. set maxIdleTime=61&sessionVariables=wait_timeout=180
2. wait for 180 sec. 
3. Try to access -> boom.

**Comment by tuaris** [comment]
Confirmed this is still occurring with 10.3.18-MariaDB-1:10.3.18+maria~jessie                                                                                                                                                        

```
[2d7a6e3a-5cd5-43bc-b160-0978f2453298] 2019-10-05T01:28:08-04:00 WARN metabase.query-processor.middleware.process-userland-query Query failure {:status :failed,
 :class java.sql.SQLNonTransientConnectionException,
 :error "(conn=491) Connection reset",
 :stacktrace
 ("org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:234)"
  "org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:165)"
  "org.mariadb.jdbc.MariaDbStatement.executeExceptionEpilogue(MariaDbStatement.java:238)"
  "org.mariadb.jdbc.MariaDbStatement.executeInternal(MariaDbStatement.java:356)"
  "org.mariadb.jdbc.MariaDbStatement.executeQuery(MariaDbStatement.java:530)"
  "org.mariadb.jdbc.MariaDbConnection.getTransactionIsolation(MariaDbConnection.java:951)"
  "com.mchange.v2.c3p0.impl.NewProxyConnection.getTransactionIsolation(NewProxyConnection.java:1278)"
  "clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:790)"
  "clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
  "clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:834)"
  "clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
  "clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:782)"
  "clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
  "--> driver.sql_jdbc.execute$do_in_transaction.invokeStatic(execute.clj:241)"
  "driver.sql_jdbc.execute$do_in_transaction.invoke(execute.clj:240)"
  "driver.sql_jdbc.execute$run_query_without_timezone.invokeStatic(execute.clj:259)"
  "driver.sql_jdbc.execute$run_query_without_timezone.invoke(execute.clj:258)"
  "driver.sql_jdbc.execute$execute_query$fn__66901.invoke(execute.clj:288)"
  "driver.sql_jdbc.execute$do_with_try_catch.invokeStatic(execute.clj:217)"
  "driver.sql_jdbc.execute$do_with_try_catch.invoke(execute.clj:211)"
  "driver.sql_jdbc.execute$execute_query.invokeStatic(execute.clj:285)"
  "driver.sql_jdbc.execute$execute_query.invoke(execute.clj:279)"
  "driver.sql_jdbc$fn__67145.invokeStatic(sql_jdbc.clj:44)"
  "driver.sql_jdbc$fn__67145.invoke(sql_jdbc.clj:43)"
  "query_processor$fn__43901$execute_query__43906$fn__43907.invoke(query_processor.clj:70)"
  "query_processor$fn__43901$execute_query__43906.invoke(query_processor.clj:64)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__34042.invoke(mbql_to_native.clj:38)"
  "query_processor.middleware.annotate$result_rows_maps__GT_vectors$fn__36286.invoke(annotate.clj:540)"
  "query_processor.middleware.annotate$add_column_info$fn__36192.invoke(annotate.clj:484)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__37213.invoke(cumulative_aggregations.clj:57)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__41011.invoke(resolve_joins.clj:184)"
  "query_processor.middleware.limit$limit$fn__37949.invoke(limit.clj:19)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__43770.invoke(results_metadata.clj:87)"
  "query_processor.middleware.format_rows$format_rows$fn__37937.invoke(format_rows.clj:26)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__34801.invoke(add_dimension_projections.clj:234)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__35420.invoke(add_source_metadata.clj:107)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__41061.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__35280.invoke(add_row_count_and_status.clj:16)"
  "query_processor.middleware.driver_specific$process_query_in_context$fn__37424.invoke(driver_specific.clj:12)"
  "query_processor.middleware.resolve_driver$resolve_driver$fn__40675.invoke(resolve_driver.clj:22)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__36615$fn__36616.invoke(bind_effective_timezone.clj:9)"
  "util.date$call_with_effective_timezone.invokeStatic(date.clj:88)"
  "util.date$call_with_effective_timezone.invoke(date.clj:77)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__36615.invoke(bind_effective_timezone.clj:8)"
  "query_processor.middleware.store$initialize_store$fn__43795$fn__43796.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__43795.invoke(store.clj:10)"
  "query_processor.middleware.async$async__GT_sync$fn__33953.invoke(async.clj:23)"
  "query_processor.middleware.async_wait$runnable$fn__36343.invoke(async_wait.clj:89)"),
 :query
 {:type "query",
  :query {:source-table 25, :joins [{:fields "all", :source-table 10, :condition ["=" ["field-id" 1279] ["joined-field" "Project" ["field-id" 109]]], :alias "Project"}]},
  :parameters [],
  :async? true,
  :middleware {:add-default-userland-constraints? true, :userland-query? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :card-id nil,
   :nested? false,
   :query-hash [36, -25, -116, -62, 4, 112, 109, -63, -58, -7, 32, -70, 58, 3, -42, 125, -77, 126, 30, -59, 7, 100, 54, -58, -1, -60, -127, -12, 115, 20, -32, -94]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :preprocessed
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :query,
  :middleware {:add-default-userland-constraints? true, :userland-query? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [36, -25, -116, -62, 4, 112, 109, -63, -58, -7, 32, -70, 58, 3, -42, 125, -77, 126, 30, -59, 7, 100, 54, -58, -1, -60, -127, -12, 115, 20, -32, -94]},
  :preprocessing-level 1,
....
 :cause {:class java.sql.SQLException, :error "Connection reset", :cause {:class java.net.SocketException, :error "Connection reset"}}}
```

**Comment by flamber** [comment]
@tuaris This, along with several other issues, might be resolved by #11018 (scheduled for 0.34.0)

**Comment by camsaul** [comment]
Same issue as #9885
